ui/app/src/components/PreviewDialog/PreviewDialogContainer.tsx

```diff
@@ -26,23 +26,32 @@ import DialogFooter from '../DialogFooter';
 import SecondaryButton from '../SecondaryButton';
 import { FormattedMessage } from 'react-intl';
 import PrimaryButton from '../PrimaryButton';
-import useDetailedItem from '../../hooks/useDetailedItem';
 import { DialogBody } from '../DialogBody';
 import { useDispatch } from 'react-redux';
 import { closePreviewDialog, showCodeEditorDialog } from '../../state/actions/dialogs';
 import { batchActions } from '../../state/actions/misc';
 import { hasEditAction, isBlobUrl } from '../../utils/content';
 import { useSelection } from '../../hooks/useSelection';
+import useItemsByPath from '../../hooks/useItemsByPath';
+import useActiveSiteId from '../../hooks/useActiveSiteId';
+import { fetchSandboxItem } from '../../state/actions/content';
 
 export function PreviewDialogContainer(props: PreviewDialogContainerProps) {
-  const { title, content, mode, url, onClose, type, mimeType, backgroundModeIndex } = props;
+  const { title, content, mode, url, path, onClose, type, mimeType, backgroundModeIndex } = props;
   const { classes, cx } = useStyles();
-  const item = useDetailedItem(url);
+  const siteId = useActiveSiteId();
+  const items = useItemsByPath();
+  const item = items?.[path];
   const dispatch = useDispatch();
   const guestBase = useSelection<string>((state) => state.env.guestBase);
-
   const [isLoading, setIsLoading] = useState(true);
 
+  useEffect(() => {
+    if (type === 'editor' && path && !items[path]) {
+      dispatch(fetchSandboxItem({ path }));
+    }
+  }, [siteId, items, path, type, dispatch]);
```

Comment:
Does this cause multiple fetch calls to be made while the thing is being fetched? You may need to also use `itemsBeingFetchedByPath` to avoid issuing multiple calls. Use network throttle test the case well.